### ISSUE TITLE
CSP nonce by default

### ISSUE THREAD

**Issue by rullzer** [nextcloud/server#10207]
In https://github.com/nextcloud/server/pull/10205 I tried to always add the CSP nonce.

As most browsers if unsafe-inline is there and a nonce will ignore the unsafe-inline. This is the CSPv3 backwards compatibility idea.

However this seems to break on Edge. As edge doesn't properly parse the nonce on external resources.

We should investigate how to enable the nonce on more browsers.

**Comment by MorrisJobke** [comment]
> We should investigate how to enable the nonce on more browsers.

Safari: #11966

**Comment by MichaIng** [comment]
@rullzer 
Not sure, if it's related, but since I updated to NC15.0.0 Beta 2, many images (various icons on all NC pages) are blocked by CSP with: `Header always set Content-Security-Policy "default-src 'self' 'unsafe-inline' 'unsafe-eval';"` (Apache)

I guess it's related to this topic, but I don't understand enough to know how to fix best. Simply removing the CSP header works, which is fine for me now, but would be great, after work on this is done/NC15 release to have e.g. some page on docs about the minimum CSP entries required to show all Nextcloud core content.

This is on Opera developer browser v58, but should be not related. I am open for testing/log reports, in case.

**Comment by rullzer** [comment]
You should not set your own csp headers. Nextcloud does this all automatically.

**Comment by MichaIng** [comment]
@rullzer 
Ah okay, makes sense then.

But if I want to set my own CSP headers globally for other/own websites, is it possible to do this without affecting the ones from Nextcloud? Perhaps `Header always set Content-Security-Policy ""` within Nextcloud vhost/location config to unset global headers first?

**Comment by szaimen** [comment]
I'd vote for closing this now that Edge was superseded by Chromium Edge.

**Comment by MichaIng** [comment]
I guess this means to re-open this issue for discussion, doesn't it?

Currently the nonce is still only used based on a whitelist:
- https://github.com/nextcloud/server/blob/215aef3/lib/private/AppFramework/Middleware/Security/CSPMiddleware.php#L74-L76
- https://github.com/nextcloud/server/blob/215aef3/lib/private/Security/CSP/ContentSecurityPolicyNonceManager.php#L80-L95

If compatibility with old Microsoft Edge is not required, then #10205 could be reattempted, switching to a blacklist. But there are a lot of browsers out there, including the mobile variants, so much to test 🙂.

**Comment by szaimen** [comment]
cc @nextcloud/security

**Comment by susnux** [comment]
The only reason to not do this would be problems on client side for login flow, so if there are no objections from @nextcloud/android or  @nextcloud/ios we can enable it by default.
At least for desktop all supported browsers support CSP nonce.

**Comment by MichaIng** [comment]
This should be the relevant support list: https://caniuse.com/contentsecuritypolicy2
It also shows the broken nonce support in old Edge. The info regarding Opera Mini is almost always wrong, as it is very outdated (2015). Current Opera Mini supports all modern browser features, just like Opera Mobile.

**Comment by tobiasKaminsky** [comment]
Can I test this somehow before we merge it?
I would like to test it on older Androids. They should have an up to date web view, but better safe than sorry.

**Comment by MichaIng** [comment]
Here is an example about how to define a nonce in an inline CSS block, and a related CSP rule: to block `unsafe-line` CSS, but only block with this particular nonce: https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy/style-src#unsafe_inline_styles

**Comment by tobiasKaminsky** [comment]
I am to n00b to have an idea how to put this in our login flow.
Sorry for that :innocent: 

Other places that clients use, which are served by served are
OnlyOffice
Collabora/NC Office
Text editor

**Comment by susnux** [comment]
What is the lowest webview we support? For all features to work with nonce we would need at least android webview 54 & safari 15.4.

If this does not work for our supported clients, how about moving to a blacklist instead of the whitelist and putting in our client UA?

**Comment by MichaIng** [comment]
> I am to n00b to have an idea how to put this in our login flow.

I was just thinking to test this outside of Nextcloud, with a single `index.html` which just serves an image with a nonce, and a related CSP rule to block everything else but `img-src 'nonce-somehash'`. A second image without nonce can serve to verify that the CSP rule really blocks everything else. If the image with the nonce is visible, the browser does not support nonce sources (in CSP rules), if it is hidden (while the control image is visible), all is good.

> What is the lowest webview we support? For all features to work with nonce we would need at least android webview 54 & safari 15.4.

Did you test this? caniuse says Safari 10 is sufficient, and Chromium 40, which should also equal webview 40. However, for Android 5, there is much newer webview available, while for Android 4.4, AFAICS, it is stuck below version 40. So if this information was true, we would require Android 5+, which does not sound alike a problem to me.

**Comment by susnux** [comment]
BTW I saw we only support Android 6+ officially  -> So this should be good :)

**Comment by susnux** [comment]
> But as we support "only" Android 7+ it might not be needed?

@tobiasKaminsky Then the docs are outdated they say 6+:
https://docs.nextcloud.com/server/latest/admin_manual/installation/system_requirements.html#mobile-apps

**Comment by nickvergessen** [comment]
At least Talk is still 6+

**Comment by MichaIng** [comment]
I mean it is of course possible that people run older webview on newer Android, theoretically. Let's see Chromium (=webview) and Android release dates:
- https://www.chromium.org/developers/calendar/#previous-release-information
- https://en.wikipedia.org/wiki/Android_version_history#Android_6.0_Marshmallow

So oldest webview possible on Android 6 is probably v46. On Android 7 probably v53. If CSP2 is all that is needed, webview 40 should be sufficient, but probably there is another feature which requires v54 as you sais @susnux ?

However, with PlayStore and all, it is pretty unlikely and a misbehaviour to run such outdated webview, when the very recent is available for all those Android versions, hence I would not see this as blocker.

For Safari it is probably a bigger problem, since Safari versions are capped on macOS/iOS versions. Safari 10 wouldn't be an issue, but for Safari 15 one requires at least macOS 10.15 and iOS 15, it seems:https://en.wikipedia.org/wiki/Safari_(web_browser)#Version_compatibility
Not sure how common/uncommon it is to run older macOS/iOS versions.

For easier testing, I created a simple test page: https://dietpi.com/nonce/

**Comment by SystemKeeper** [comment]
Would it be a lot of work to put the changes in a server PR so we can test against different iOS/Android versions?

**Comment by MichaIng** [comment]
@SystemKeeper 
Test it with the website I posted above: https://dietpi.com/nonce/
No need to wait for a PR.

**Comment by susnux** [comment]
> Would it be a lot of work to put the changes in a server PR so we can test against different iOS/Android versions?

Try this one: https://github.com/nextcloud/server/pull/44412
